corehq/apps/app_manager/views/utils.py

```diff
@@ -623,10 +657,37 @@ def _is_duplicate_endpoint_id(new_id, old_id, app):
         all_endpoint_ids.append(module.session_endpoint_id)
         for form in module.get_suite_forms():
             all_endpoint_ids.append(form.session_endpoint_id)
+        if module.module_type == "shadow":
+            for m in module.form_session_endpoint_mapping:
+                all_endpoint_ids.append(m.session_endpoint_id)
 
     return new_id in all_endpoint_ids
 
 
+def _duplicate_endpoint_ids(new_session_endpoint_id, new_form_session_endpoint_ids, module_id, app):
```

Comment:
Any way to combine this function with the one above it?

Reply:
do you mean with `set_shadow_module_and_form_session_endpoint`?

Reply:
I meant with `_is_duplicate_endpoint_id` - both go through the app and pull out all endpoint IDs, though they're a little different, was just thinking it'd be nice to combine them if possible, but I'm not certain whether it'd come out better or worse in the end.